board: mimxrt1170_evk

TimTTP opened the pull request:
Hi there,

I have been using this board for some other work and wanted to test out Zephyr with it, seeing as it is a supported board. However, I do not have a J-Link debugger to hand so the flash and debugging instructions weren't working. Looking more closely at the instructions, I was able to use `pyocd` to flash and debug the elf file directly. The MIMXRT1176 processor on this board is now supported by `pyocd` where perhaps it wasn't before: https://pyocd.io/docs/builtin-targets.html, https://github.com/pyocd/pyOCD/blob/main/pyocd/target/builtin/target_MIMXRT1176xxxxx.py.

It should also be noted that most of the MIMXRT1xxx family now appear to be supported by `pyocd` yet the docs all include similar language about flash support not being ready yet. I don't have any of the other boards/processors so cannot confirm this.

Seeing as the `pyocd` runner doesn't require any extra hardware, I think it is easier for getting started and therefore makes a better default. So I have reordered the includes in `board.cmake` which seems to have done the trick, although I couldn't see any explicit documentation on this. I will include at the bottom of this message the terminal output I see when running the commands with this change.

This is my first Zephyr contribution, so I have tried to make minimal changes and keep docs information in line with what is already there. Please feel free to refer me to other examples where `pyocd` and `J-Link` are supported so I can copy their style/information. Or if I have missed anything with regards to contributing guidelines.

Thanks 😃 

Tim

P.S Terminal output following this change:

```bash
>west --verbose flash
-- west flash: rebuilding
cmake version 3.26.3 is OK; minimum version is 3.13.1
Running CMake: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\tim\zephyr_project\build'
ninja: no work to do.
-- west flash: using runner pyocd
WARNING: runners.pyocd: hex file (None) does not exist; falling back on .bin (C:\Users\tim\zephyr_project\build\zephyr\zephyr.bin). Consider enabling CONFIG_BUILD_OUTPUT_HEX.
-- runners.pyocd: Flashing file: C:\Users\tim\zephyr_project\build\zephyr\zephyr.bin
runners.pyocd: pyocd flash -e sector -a 0x30000000 -t mimxrt1170_cm7 'C:\Users\tim\zephyr_project\build\zephyr\zephyr.bin'
0001998 I Loading C:\Users\tim\zephyr_project\build\zephyr\zephyr.bin at 0x30000000 [load_cmd]
[==================================================] 100%
0008938 I Erased 36864 bytes (9 sectors), programmed 35584 bytes (139 pages), skipped 4096 bytes (16 pages) at 5.65 kB/s [loader]

>west --verbose debug
-- west debug: rebuilding
cmake version 3.26.3 is OK; minimum version is 3.13.1
Running CMake: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\tim\zephyr_project\build'
ninja: no work to do.
-- west debug: using runner pyocd
-- runners.pyocd: pyOCD GDB server running on port 3333
runners.pyocd: pyocd gdbserver -p 3333 -T 4444 -t mimxrt1170_cm7
runners.pyocd: C:/Users/tim/zephyr-sdk-0.16.1/arm-zephyr-eabi/bin/arm-zephyr-eabi-gdb.exe 'C:\Users\tim\zephyr_project\build\zephyr\zephyr.elf' -ex 'target remote :3333' -ex 'monitor halt' -ex 'monitor reset' -ex load
GNU gdb (Zephyr SDK 0.16.1) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-host_w64-mingw32 --target=arm-zephyr-eabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://github.com/zephyrproject-rtos/sdk-ng/issues>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from C:\Users\tim\zephyr_project\build\zephyr\zephyr.elf...
0001337 I Target type is mimxrt1170_cm7 [board]
0001378 I DP IDR = 0x6ba02477 (v2 rev6) [dap]
0001443 I AHB-AP#0 IDR = 0x84770001 (AHB-AP var0 rev8) [discovery]
0001490 I AHB-AP#1 IDR = 0x24770011 (AHB-AP var1 rev2) [discovery]
0001507 I APB-AP#2 IDR = 0x54770002 (APB-AP var0 rev5) [discovery]
0001560 I AHB-AP#0 Class 0x1 ROM table #0 @ 0xe00fd000 (designer=00e:NXP part=88c) [rom_table]
0001567 I [0]<e00fe000:ROM class=1 designer=43b:Arm part=4c8> [rom_table]
0001570 I   AHB-AP#0 Class 0x1 ROM table #1 @ 0xe00fe000 (designer=43b:Arm part=4c8) [rom_table]
0001580 I   [0]<e00ff000:ROM class=1 designer=43b:Arm part=4c7> [rom_table]
0001581 I     AHB-AP#0 Class 0x1 ROM table #2 @ 0xe00ff000 (designer=43b:Arm part=4c7) [rom_table]
0001640 I     [0]<e000e000:SCS v7-M class=14 designer=43b:Arm part=00c> [rom_table]
0001644 I     [1]<e0001000:DWT v7-M class=14 designer=43b:Arm part=002> [rom_table]
0001648 I     [2]<e0002000:FPB v7-M class=14 designer=43b:Arm part=00e> [rom_table]
0001660 I     [3]<e0000000:ITM v7-M class=14 designer=43b:Arm part=001> [rom_table]
0001727 I   [1]<e0041000:ETM M7 class=9 designer=43b:Arm part=975 devtype=13 archid=4a13 devid=0:0:0> [rom_table]
0001732 I   [2]<e0042000:CTI CS-400 class=9 designer=43b:Arm part=906 devtype=14 archid=0000 devid=40800:0:0> [rom_table]
0001738 I [1]<e0043000:Trace Funnel CS-400 class=9 designer=43b:Arm part=908 devtype=12 archid=0000 devid=28:0:0> [rom_table]
0001756 I AHB-AP#1 Class 0x1 ROM table #0 @ 0xe00ff000 (designer=43b:Arm part=4c4) [rom_table]
0001830 I [0]<e000e000:SCS v7-M class=14 designer=43b:Arm part=00c> [rom_table]
0001834 I [1]<e0001000:DWT v7-M class=14 designer=43b:Arm part=002> [rom_table]
0001840 I [2]<e0002000:FPB v7-M class=14 designer=43b:Arm part=003> [rom_table]
0001844 I [3]<e0000000:ITM v7-M class=14 designer=43b:Arm part=001> [rom_table]
0001856 I [5]<e0041000:ETM M4 class=9 designer=43b:Arm part=925 devtype=13 archid=0000 devid=0:0:0> [rom_table]
0001862 I [7]<e0043000:Trace Funnel CS-400 class=9 designer=43b:Arm part=908 devtype=12 archid=0000 devid=28:0:0> [rom_table]
0001868 I [8]<e0042000:CTI CS-400 class=9 designer=43b:Arm part=906 devtype=14 archid=0000 devid=40800:0:0> [rom_table]
0001878 I CPU core #0 is Cortex-M7 r1p2 [cortex_m]
0001885 I FPU present: FPv5-D16-M [cortex_m]
0001982 I 4 hardware watchpoints [dwt]
0002016 I 8 hardware breakpoints, 1 literal comparators [fpb]
0002143 I 4 hardware watchpoints [dwt]
0002175 I 6 hardware breakpoints, 4 literal comparators [fpb]
0002400 I Semihost server started on port 4444 (core 0) [server]
Remote debugging using :3333
0002710 I GDB server started on port 3333 (core 0) [gdbserver]
0002713 I Client connected to port 3333! [gdbserver]
0x002230e8 in ?? ()
0002745 I Attempting to load RTOS plugins [gdbserver]
Successfully halted device
Resetting target
Loading section rom_start, size 0x23a8 lma 0x30000000
Loading section text, size 0x45ac lma 0x300023a8
Loading section .ARM.exidx, size 0x8 lma 0x30006954
Loading section initlevel, size 0xa8 lma 0x3000695c
Loading section device_area, size 0x154 lma 0x30006a04
Loading section sw_isr_table, size 0x6d0 lma 0x30006b58
Loading section log_const_area, size 0x28 lma 0x30007228
Loading section rodata, size 0x2828 lma 0x30007250
Loading section datas, size 0x10 lma 0x30009a78
Loading section device_states, size 0x22 lma 0x30009a88
Loading section .last_section, size 0x4 lma 0x30009aaa
[==================================================] 100%
0005688 I Erased 0 bytes (0 sectors), programmed 0 bytes (0 pages), skipped 39680 bytes (155 pages) at 13.87 kB/s [loader]
--Type <RET> for more, q to quit, c to continue without paging
```

DerekSnell reviewed (changes requested):
Hi @TimTTP ,
I appreciate your contribution, and trying to improve the debug experience with this board.  I understand the reasoning for your change.  Unfortunately, you stumbled into a topic that NXP is actively working on to change for all our boards.  And your PR conflicts with our plan.

NXP recently released a debug tool called LinkServer, which supports debug probes based on CMSIS-DAP.  We recently merged our first LinkServer PR https://github.com/zephyrproject-rtos/zephyr/pull/58015 adding LinkServer support to Zephyr for a few boards.  Our plan is to finish updating all the NXP boards for LinkServer support, and make LinkServer the default runner for these boards.  @zanderzhng has already jumped ahead of NXP, and submitted https://github.com/zephyrproject-rtos/zephyr/pull/61149 for this same RT1170 EVK.

LinkServer will support the on-board debug probes on our boards by default, because they are programmed with a CMSIS-DAP firmware.  And LinkServer provides all Zephyr users with a free debug tool option, both using West/GDB or NXP's [MCUXpresso extension for VS Code](https://github.com/nxp-mcuxpresso/vscode-for-mcux/wiki/Working-with-Zephyr).  And LinkServer is well supported by NXP for our SOCs, boards, external memories, and our customers.  Therefore, we plan to update our board doc pages with the primary debug options of LinkServer and JLink, with LinkServer being default.  

I am sorry our LinkServer plan was not further along before you put the effort in for this PR.  And I am sorry to block it with the good effort you put into this, but we do not want pyOCD to be the default runner for the NXP boards.  I suggest you wait, and let NXP push our LinkServer changes out, and update the board doc pages.  And we will welcome your feedback in those updates.

I also see you suggested changes to the [Flashing section](https://docs.zephyrproject.org/latest/boards/arm/mimxrt1170_evk/doc/index.html#flashing) of the doc page.  I believe the reason the steps are written like this is to enable flash programming even when the external flash is corrupted.  This happens frequently to users during development if the programming is interrupted or fails.  And sometimes it locks out debug access.  By booting with SW1 set to 0001b, it forces the boot ROM into Serial Downloader mode, and does not attempt to boot from the flash.  Then the flash can be programmed.  Your suggested steps will work most of the time.  But when that does not work, we want users of this board to know about the process that can recover the flash.

Thanks again for your effort here.

TimTTP commented:
Hi @DerekSnell, thanks for reviewing this PR. Glad to hear that NXP are working hard to support Zephyr 👍 I'm happy to strip back most of the changes in light of these updates, but I still think it would be useful to correct the following line on the docs page:

```
however the [pyOCD Debug Host Tools](https://docs.zephyrproject.org/latest/develop/flash_debug/host-tools.html#pyocd-debug-host-tools) do not yet support programming the external flashes on this board
```

As above, `pyocd` does now support programming the flashes on this board, I just tested it this morning in the `0001b` setting to double check. The line above is included verbatim in most of the mimxrt docs (see below), so if we can come up with a correction for this board, it could be verified by NXP and applied to all of the other board docs as well.

This would not interfere with either of the PRs you have referred to above.

```
$ grep -r 'however the :ref:`pyocd-debug-host-tools` do not yet support programming the' --include=*.rst boards/
boards/arm/mimxrt1010_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1015_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1020_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1024_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1040_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1050_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1060_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1062_fmurt6/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1064_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mimxrt1160_evk/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
boards/arm/mm_swiftio/doc/index.rst:however the :ref:`pyocd-debug-host-tools` do not yet support programming the
```

TimTTP commented:
I've added another commit now which removes changes to `board.cmake` and some of the other changes to the docs, but adds a small section detailing how to use `pyOCD`. I have tried to keep this in line with language in other docs such as Nucleo h563zi which also includes notes on `pyocd` as an alternative flash/debug runner.

As above, if someone with access to all of the NXP MIMXRT boards can confirm this is working on them as well then we can potentially apply the same patch to all of those docs without interfering with any of the LinkServer work 👍

Please let me know if you think this is suitable.

Thanks,

Tim

DerekSnell reviewed (changes requested):
Hi @TimTTP ,
Again, I do appreciate your efforts to improve this doc page.  But NXP has decided once LinkServer support is added for a board, we will remove pyOCD references from the board doc pages.  Those pages will document using the default LinkServer runner, and using JLink.  pyOCD will be left as a runner in the `board.cmake` files.  But our board docs will be updated to encourage the well-supported options of LinkServer and JLink.  

We are currently working on the LinkServer PR for this specific board, and I will be updating the doc page in that PR.  So again, I am sorry to block your PR.  I do look forward to your future contributions that I can help with, instead of being the bad guy :)

And for future reference with other PRs, with your PR updates this PR now has 2 commits, and 1 is out dated.  That would require a change.  In this case a single commit would be good.  Thanks

DerekSnell commented:
FYI @TimTTP , we submitted our PR [Enable LinkServer support on NXP RT1170](https://github.com/zephyrproject-rtos/zephyr/pull/62229), which also updates the RT1170 board doc page for details using the default CMSIS-DAP on-board debugger firmware with LinkServer.  Let us know if you have any feedback.  Thanks again for your efforts to improve the debug experience and documentation for this board.

TimTTP commented:
Hi @DerekSnell, sorry for my late reply. Good to see you got the LinkServer change completed, I'm sure that will be helpful.

I am a bit short on time at the moment but would like to make it clearer that `pyocd` is an option for developers who are not familiar with Zephyr (like me).

I will make relevant changes next week.

Thanks,

Tim

zanderzhng commented:
@TimTTP @DerekSnell Sorry for this late reply, I was on a two-week business trip and just came back to my workstation today and had a chance to check these messages. I will try to walk through and understand what's going on asap. If you need anything from me urgently, please let me know.

DerekSnell commented:
> @TimTTP @DerekSnell Sorry for this late reply, I was on a two-week business trip and just came back to my workstation today and had a chance to check these messages. I will try to walk through and understand what's going on asap. If you need anything from me urgently, please let me know.

Hi @zanderzhng ,
There is nothing urgent here.  Our PR [Enable LinkServer support on NXP RT1170](https://github.com/zephyrproject-rtos/zephyr/pull/62229) has already merged.  I was just including you because your PR https://github.com/zephyrproject-rtos/zephyr/pull/61149 was related to this one.  Thanks

dleach02 commented:
@TimTTP, can we close this PR given the feedback from @DerekSnell and the recently merged LinkServer PR?

TimTTP commented:
Hi @dleach02, sorry for leaving this open for so long. I still think a reference to pyOCD would be really helpful as I struggled to find this as a beginner with Zephyr. Although the NXP tools are great, they might not be suitable for everyone and don't work straight out of the box with Zephyr. If this is agreed with then I'm happy to suggest another change but otherwise it can be closed, yes.

TimTTP committed:
board: mimxrt1170_evk Add pyOCD runner to docs

Added a short paragraph explaining how to use pyOCD to flash and debug this
board, as well as a disclaimer requested by NXP

Signed-off-by: Tim Guite <tim.guite@ttp.com>

TimTTP commented:
I have now made what I believe is a minimal change to the docs that makes it clear that pyOCD is a working option for this board. As above, pyOCD is a straightforward option for beginners as it is installed as part of the setup process. Knowing this would have helped me a lot when learning how to use Zephyr with this board.

@DerekSnell I understand your note above that NXP is encouraging the use of LinkServer and JLink. If you can give the commit as is it is now, and my comments, a short review I would be grateful. If you decide that it is still not suitable then please close this PR. Thanks for working on the Zephyr support for this board in the first place :)

DerekSnell commented:
Hi @TimTTP ,
I appreciate your commitment to keeping the pyOCD option in this doc page.  I will review your request internally at NXP, and get back to you.  Thank you

TimTTP commented:
Thanks @DerekSnell, look forward to hearing from you 👍

DerekSnell commented on boards/arm/mimxrt1170_evk/doc/index.rst:
```diff
@@ -349,6 +349,9 @@ Use the ``-r linkserver`` option with West to use the LinkServer runner.
 
+Alternatively, pyOCD can be used to flash and debug the board by using the
+``-r pyocd`` option with West. pyOCD is installed when you complete the
+:ref:`gs_python_deps` step in the Getting Started Guide.
```
```suggestion
:ref:`gs_python_deps` step in the Getting Started Guide.  The runners supported by NXP are LinkServer and JLink.  pyOCD is another potential option, but NXP does not test or support the pyOCD runner.
```

TimTTP replied:
Thanks Derek, added that now 👍

DerekSnell reviewed (changes requested):
Hi @TimTTP ,
Thanks to your commitment to pyOCD, NXP has decided to leave details of the pyOCD runner support in our board doc pages.  However, NXP requires this disclaimer to be added about pyOCD.  We will also add this disclaimer to all NXP boards with both pyOCD and LinkServer support.

Once you add this disclaimer, I will approve this PR.  Thanks for all your efforts with this.

DerekSnell reviewed (approved):
Hi @TimTTP ,
Thank you for your commitment to make this change.  Best regards

TimTTP commented:
Thanks for your help, @DerekSnell, @butok, @dleach02 😄
